src/freenet/pluginmanager/PluginManager.java

```diff
@@ -1183,266 +1068,246 @@ public File getPluginFilename(String pluginName) {
 	 *
 	 * @param name
 	 *            The specification of the plugin
-	 * @param ignoreOld If true, ignore any old cached copies of the plugin,
-	 * and download a new version anyway. This is especially important on Windows,
-	 * where we will not usually be able to delete the file after determining
-	 * that it is too old. 
+	 * @param alwaysDownload If true, always download a new version anyway.
+	 * This is especially important on Windows, where we will not usually be
+	 * able to delete the file after determining that it is too old.
 	 * @return An instanciated object of the plugin
 	 * @throws PluginNotFoundException
 	 *             If anything goes wrong.
+	 * @throws PluginAlreadyLoaded if the plugin is already loaded
 	 */
-	private FredPlugin loadPlugin(PluginDownLoader<?> pdl, String name, PluginProgress progress, boolean ignoreOld) throws PluginNotFoundException {
+	private FredPlugin loadPlugin(PluginDownLoader<?> pdl, String name, PluginProgress progress, boolean alwaysDownload) throws PluginNotFoundException, PluginAlreadyLoaded {
 
 		pdl.setSource(name);
 
-		/* check for plugin directory. */
-		File pluginDirectory = node.getPluginDir();
-		if((pluginDirectory.exists() && !pluginDirectory.isDirectory()) || (!pluginDirectory.exists() && !pluginDirectory.mkdirs())) {
-			Logger.error(this, "could not create plugin directory");
-			throw new PluginNotFoundException("could not create plugin directory");
-		}
+		File pluginDirectory = getPluginDirectory();
 
 		/* get plugin filename. */
 		String filename = pdl.getPluginName(name);
-		boolean pluginIsLocal = pdl instanceof PluginDownLoaderFile;
-		File pluginFile = new File(pluginDirectory, filename + "-" + System.currentTimeMillis());
-
-		/* check for previous instances and delete them. */
-		File[] filesInPluginDirectory = getPreviousInstances(pluginDirectory, filename);
-		boolean first = true;
-		for (File cachedFile : filesInPluginDirectory) {
-			if (first && !pluginIsLocal && !ignoreOld) {
-				first = false;
-				pluginFile = new File(pluginDirectory, cachedFile.getName());
-				continue;
-			}
-			first = false;
-			cachedFile.delete();
-		}
+		File pluginFile = getTargetFileForPluginDownload(pluginDirectory, filename, !pdl.isCachingProhibited() && !alwaysDownload);
 
 		boolean downloaded = false;
 		/* check if file needs to be downloaded. */
 		if(logMINOR)
 			Logger.minor(this, "plugin file " + pluginFile.getAbsolutePath() + " exists: " + pluginFile.exists()+" downloader "+pdl+" name "+name);
 		int RETRIES = 5;
-		for(int i = 0; i < RETRIES; i++) {
-			if(!pluginFile.exists() || pluginFile.length() == 0)
+		for (int i = 0; i < RETRIES; i++) {
+			if (!pluginFile.exists() || pluginFile.length() == 0) {
 				try {
 					downloaded = true;
```

Comment:
Why is this flag set here? It's clearly not downloaded at this point and the flag will never be reset to `false`.